pandas/tseries/index.py

```diff
@@ -329,10 +329,14 @@ def __new__(cls, data=None,
                     subarr = tslib.cast_to_nanoseconds(data)
                 else:
                     subarr = data
-        elif data.dtype == _INT64_DTYPE:
+        elif is_integer_dtype(data.dtype):
             if isinstance(data, Int64Index):
                 raise TypeError('cannot convert Int64Index->DatetimeIndex')
+            if data.dtype != _INT64_DTYPE:
+                data = data.astype(np.int64)
             subarr = data.view(_NS_DTYPE)
+        else:
+            data = subarr
```

Comment:
What is this for? (Or should it be the way around?) In any case this line is not covered by the tests I think